eth/utils/bls.py

```diff
@@ -208,16 +211,19 @@ def verify_multiple(pubkeys: Sequence[int],
             )
         )
 
-    o = FQ12([1] + [0] * 11)
-    for m_pubs in set(messages):
-        # aggregate the pubs
-        group_pub = Z1
-        for i in range(len_msgs):
-            if messages[i] == m_pubs:
-                group_pub = add(group_pub, decompress_G1(pubkeys[i]))
-
-        o *= pairing(hash_to_G2(m_pubs, domain), group_pub, False)
-    o *= pairing(decompress_G2(signature), neg(G1), False)
-
-    final_exponentiation = final_exponentiate(o)
-    return final_exponentiation == FQ12.one()
+    try:
+        o = FQ12([1] + [0] * 11)
+        for m_pubs in set(messages):
+            # aggregate the pubs
+            group_pub = Z1
+            for i in range(len_msgs):
```

Comment:
This looks like you could do `for pubkey, message in zip(pubkeys, messages):` which removes the indexing.

Alternatively, I think the following is a functional equivalent implementation.

```python
groups = (
    pairing(hash_to_G2(m_pubs, domain), reduce(add, (decompress_G1(pubkey), Z1)), False)
    for (pubkey, message) 
    in zip(pubkeys, messages) if message == m_pubs
    for m_pubs in set(messages)
)
final_exponentiation = final_exponentiate(
    reduce(
        operator.mul, 
        groups,
        FQ12([1] + [0] * 11),
    ) * pairing(decompress_G2(signature), neg(G1), False)
)
return final_exponentiation == FQ12.one()
```

Also, there are a few constants in there that could be lifted out such as:

- `FQ12.one()`
- `neg(G1)`

The above is just hand-written code without being tested so you should verify if you choose to switch.  I *think* the above code would be more performant... but that's just intuition.